fish-rust/src/env_dispatch.rs

```diff
@@ -0,0 +1,796 @@
+    use crate::reader::CursorSelectionMode;
+
+    let inclusive = vars
+        .get(L!("fish_cursor_selection_mode"))
+        .as_ref()
+        .map(|v| v.as_string())
+        .map(|v| v == L!("inclusive"))
+        .unwrap_or(false);
+    let mode = if inclusive {
+        CursorSelectionMode::Inclusive
+    } else {
+        CursorSelectionMode::Exclusive
+    };
+
+    let mode = mode as u8;
+    crate::ffi::reader_change_cursor_selection_mode(mode);
+}
+
+fn handle_autosuggestion_change(vars: &dyn Environment) {
+    // TODO: This was a call to reader_set_autosuggestion_enabled(vars) and
+    // reader::check_autosuggestion_enabled() should be private to the `reader` module.
+    crate::ffi::reader_set_autosuggestion_enabled_ffi(crate::reader::check_autosuggestion_enabled(
+        vars,
+    ));
+}
+
+fn handle_function_path_change(_: &dyn Environment) {
+    crate::ffi::function_invalidate_path();
+}
+
+fn handle_complete_path_change(_: &dyn Environment) {
+    crate::ffi::complete_invalidate_path();
+}
+
+fn handle_tz_change(var_name: &wstr, vars: &dyn Environment) {
+    handle_timezone(var_name, vars);
+}
+
+fn handle_locale_change(vars: &dyn Environment) {
+    init_locale(vars);
+    // We need to re-guess emoji width because the locale might have changed to a multibyte one.
+    guess_emoji_width(vars);
+}
+
+fn handle_curses_change(vars: &dyn Environment) {
+    guess_emoji_width(vars);
+    init_curses(vars);
+}
+
+fn handle_fish_use_posix_spawn_change(vars: &dyn Environment) {
+    // Note that if the variable is missing or empty we default to true (if allowed).
+    if !allow_use_posix_spawn() {
+        USE_POSIX_SPAWN.store(false, Ordering::Relaxed);
+    } else if let Some(var) = vars.get(L!("fish_use_posix_spawn")) {
+        let use_posix_spawn =
+            var.is_empty() || crate::wcstringutil::bool_from_string(&var.as_string());
+        USE_POSIX_SPAWN.store(use_posix_spawn, Ordering::Relaxed);
+    } else {
+        USE_POSIX_SPAWN.store(true, Ordering::Relaxed);
+    }
+}
+
+/// Allow the user to override the limits on how much data the `read` command will process. This is
+/// primarily intended for testing, but could also be used directly by users in special situations.
+fn handle_read_limit_change(vars: &dyn Environment) {
+    let read_byte_limit = vars
+        .get_unless_empty(L!("fish_read_limit"))
+        .map(|v| v.as_string())
+        .and_then(|v| match v.to_string().parse() {
+            Ok(v) => Some(v),
+            Err(_) => {
+                // XXX: In the C++ code, this warning wasn't behind an "is_interactive_session()"
+                // check.
+                if is_interactive_session() {
+                    FLOGF!(warning, "Ignoring invalid $fish_read_limit");
+                }
+                None
+            }
+        })
+        .unwrap_or(DEFAULT_READ_BYTE_LIMIT);
+    READ_BYTE_LIMIT.store(read_byte_limit, Ordering::Relaxed);
```

Comment:
The C++ code didn't update READ_BYTE_LIMIT in the error case

Reply:
Hmmm, good catch - if the read byte limit was previously overridden with a valid value then this would reset it to the default. But surely resetting it to the default in case of an invalid value makes more sense than preserving the prior value? After all, the user *did* intend to change it? We do warn about the invalid value, so the user would have a chance to try again if they needed/wanted to, no?

Reply:
We should just do what we did before; no reason to change behavior as part of the port.